src/flow/assembly_lmh_old.hh

```diff
@@ -28,598 +47,688 @@
+};
+
+template <unsigned int dim>
+class MHMatrixAssemblyLMH : public AssemblyBase<dim>
+{
+public:
+    typedef typename DarcyLMH::EqFields EqFields;
+    typedef typename DarcyLMH::EqData EqData;
+
+    DECLARE_EXCEPTION( ExcBCNotSupported, << "BC type not supported.\n" );
+
+    static constexpr const char * name() { return "MHMatrixAssemblyLMH"; }
+
+    /// Constructor.
+    MHMatrixAssemblyLMH(EqFields *eq_fields, EqData *eq_data)
+    : AssemblyBase<dim>(0), eq_fields_(eq_fields), eq_data_(eq_data), quad_rt_(dim, 2) {
+        this->active_integrals_ = (ActiveIntegrals::bulk | ActiveIntegrals::coupling | ActiveIntegrals::boundary);
+        this->used_fields_ += eq_fields_->cross_section;
+        this->used_fields_ += eq_fields_->conductivity;
+        this->used_fields_ += eq_fields_->anisotropy;
+        this->used_fields_ += eq_fields_->sigma;
+        this->used_fields_ += eq_fields_->water_source_density;
+        this->used_fields_ += eq_fields_->extra_source;
+        this->used_fields_ += eq_fields_->storativity;
+        this->used_fields_ += eq_fields_->extra_storativity;
+        this->used_fields_ += eq_fields_->bc_type;
+        this->used_fields_ += eq_fields_->bc_pressure;
+        this->used_fields_ += eq_fields_->bc_flux;
+        this->used_fields_ += eq_fields_->bc_robin_sigma;
+        this->used_fields_ += eq_fields_->bc_switch_pressure;
     }
 
-    /** Loads the local system from a map: element index -> LocalSystem,
-     * if it exits, or if the full solution is not yet reconstructed,
-     * and reconstructs the full solution on the element.
-     * Currently used only for seepage BC.
-     */
-    void load_local_system(const DHCellAccessor& dh_cell)
-    {
-        // do this only once per element
-        if(bc_fluxes_reconstruted)
-            return;
+    /// Destructor.
+    virtual ~MHMatrixAssemblyLMH() {}
 
-        // possibly find the corresponding local system
-        auto ls = ad_->seepage_bc_systems.find(dh_cell.elm_idx());
-        if (ls != ad_->seepage_bc_systems.end())
-        {
-            arma::vec schur_solution = ad_->p_edge_solution.get_subvec(loc_schur_.row_dofs);            
-            // reconstruct the velocity and pressure
-            ls->second.reconstruct_solution_schur(schur_offset_, schur_solution, reconstructed_solution_);
+    /// Initialize auxiliary vectors and other data members
+    void initialize(ElementCacheMap *element_cache_map) {
+        //this->balance_ = eq_data_->balance_;
+        this->element_cache_map_ = element_cache_map;
 
-            postprocess_velocity(dh_cell, reconstructed_solution_);
+        fe_ = std::make_shared< FE_P_disc<dim> >(0);
+        fe_values_side_.initialize(*this->quad_low_, *fe_, update_normal_vectors);
 
-            bc_fluxes_reconstruted = true;
-        }
+        fe_values_.initialize(quad_rt_, fe_rt_, update_values | update_JxW_values | update_quadrature_points);
+
+        ndofs_ = fe_values_.n_dofs();
```

Comment:
Rather extract these as local variables where used. No performance hit as these are inlined anyway. Try to keep shared variables to a minimum.